File: LABS/lab3/E.java
import java.util.*;

class E {
    public static void main(String[] args) {
        System.out.println(solve());
    }

    public static int solve() {
        Scanner scanner = new Scanner(System.in);
        int n = scanner.nextInt();
        int m = scanner.nextInt();

        // Create adjacency list
        List<List<Map.Entry<Integer, Integer>>> adjList = new ArrayList<>();
        for (int i = 0; i <= n; i++) {
            adjList.add(new ArrayList<>());
        }

        for (int i = 0; i < m; i++) {
            int x = scanner.nextInt();
            int y = scanner.nextInt();
            int t = scanner.nextInt();
            adjList.get(x).add(new AbstractMap.SimpleEntry<>(y, t));
            adjList.get(y).add(new AbstractMap.SimpleEntry<>(x, t));
        }

        // Store traffic jams
        List<List<Integer>> trafficJams = new ArrayList<>();
        for (int i = 0; i < n; i++) {
            List<Integer> jamTimes = new ArrayList<>();
            for (int j = 0; j < n; j++) {
                jamTimes.add(scanner.nextInt());
            }
            trafficJams.add(jamTimes);
        }

        int[] distances = new int[n+1];
        Arrays.fill(distances, Integer.MAX_VALUE);
        distances[1] = 0;

        PriorityQueue<Map.Entry<Integer, Integer>> priorityQueue = new PriorityQueue<>((a, b) -> a.getValue() - b.getValue());
        priorityQueue.add(new AbstractMap.SimpleEntry<>(0, 1));
        Set<Integer> visited = new HashSet<>();

        while (!priorityQueue.isEmpty()) {
            Map.Entry<Integer, Integer> current = priorityQueue.poll();
            int currentTime = current.getKey();
            int currentIntersection = current.getValue();

            if (visited.contains(currentIntersection)) {
                continue;
            }

            visited.add(currentIntersection);

            if (currentIntersection == n) {
                return currentTime;
            }

            if (currentIntersection != 1 && trafficJams.get(currentIntersection - 1).contains(currentTime)) {
                currentTime += 1;
            }

            for (Map.Entry<Integer, Integer> neighbor : adjList.get(currentIntersection)) {
                int neighborIntersection = neighbor.getKey();
                int travelTime = neighbor.getValue();
                int newTime = currentTime + travelTime;

                if (newTime < distances[neighborIntersection]) {
                    distances[neighborIntersection] = newTime;
                    priorityQueue.add(new AbstractMap.SimpleEntry<>(newTime, neighborIntersection));
                }
            }
        }

        return -1;
    }
}

File: LABS/lab3/E.py
import heapq

def solve():
    n, m = map(int, input().split())

    
    adj_list = [[] for _ in range(n+1)]
    for _ in range(m):
        x, y, t = map(int, input().split())
        adj_list[x].append((y, t))
        adj_list[y].append((x, t))

    
    traffic_jams = []
    for _ in range(n):
        jam_times = list(map(int, input().split()))
        traffic_jams.append(jam_times)

    dis = [float('inf')] * (n+1)
    dis[1] = 0

    priority_queue = [(0, 1)] 
    visited = set()

    while priority_queue:
        ctime, cinter = heapq.heappop(priority_queue)

        if cinter in visited:
            continue

        visited.add(cinter)

        if cinter == n:
            return ctime

        if cinter != 1 and ctime in traffic_jams[cinter - 1]:
            ctime += 1

        for nbr, travel_time in adj_list[cinter]:
            new_time = ctime + travel_time
            if new_time < dis[nbr]:
                dis[nbr] = new_time
                heapq.heappush(priority_queue, (new_time, nbr))

    return -1

print(solve())
